src/coreclr/jit/simd.cpp

```diff
@@ -472,73 +472,31 @@ void Compiler::setLclRelatedToSIMDIntrinsic(GenTree* tree)
-    {
-        if (op1ObjRef->OperGet() != op2ObjRef->OperGet())
-        {
-            break;
-        }
-
-        if ((op1ObjRef->OperIs(GT_LCL_VAR) || op1ObjRef->IsLclVarAddr()) &&
-            (op1ObjRef->AsLclVarCommon()->GetLclNum() == op2ObjRef->AsLclVarCommon()->GetLclNum()))
-        {
-            return true;
-        }
-
-        if (op1ObjRef->OperIs(GT_FIELD) && (op1ObjRef->AsField()->gtFldHnd == op2ObjRef->AsField()->gtFldHnd))
-        {
-            op1ObjRef = op1ObjRef->AsField()->GetFldObj();
-            op2ObjRef = op2ObjRef->AsField()->GetFldObj();
-            continue;
-        }
-        else
-        {
-            break;
-        }
-    }
-
-    return false;
-}
-
 //----------------------------------------------------------------------
-// Check whether two field are contiguous
+// areFieldsContiguous: Check whether two fields are contiguous.
+//
 // Arguments:
-//      first - GenTree*. The Type of the node should be TYP_FLOAT
-//      second - GenTree*. The Type of the node should be TYP_FLOAT
+//      op1 - The first field indirection
+//      op2 - The second field indirection
+//
 // Return Value:
-//      if the first field is located before second field, and they are located contiguously,
-//      then return true. Otherwise, return false.
-
-bool Compiler::areFieldsContiguous(GenTree* first, GenTree* second)
+//      If the first field is located before second field, and they are
+//      located contiguously, then return true. Otherwise, return false.
+//
+bool Compiler::areFieldsContiguous(GenTreeIndir* op1, GenTreeIndir* op2)
 {
-    assert(first->OperGet() == GT_FIELD);
-    assert(second->OperGet() == GT_FIELD);
-    assert(first->gtType == TYP_FLOAT);
-    assert(second->gtType == TYP_FLOAT);
-
-    var_types firstFieldType  = first->gtType;
-    var_types secondFieldType = second->gtType;
-
-    unsigned firstFieldEndOffset = first->AsField()->gtFldOffset + genTypeSize(firstFieldType);
-    unsigned secondFieldOffset   = second->AsField()->gtFldOffset;
-    if (firstFieldEndOffset == secondFieldOffset && firstFieldType == secondFieldType &&
-        areFieldsParentsLocatedSame(first, second))
+    assert(op1->OperIs(GT_IND) && op2->OperIs(GT_IND));
+    // TODO-1stClassStructs: delete once IND<struct> nodes are no more.
+    assert(!op1->TypeIs(TYP_STRUCT) && !op2->TypeIs(TYP_STRUCT));
+
+    var_types         op1Type      = op1->TypeGet();
+    var_types         op2Type      = op2->TypeGet();
+    GenTreeFieldAddr* op1Addr      = op1->Addr()->AsFieldAddr();
+    GenTreeFieldAddr* op2Addr      = op2->Addr()->AsFieldAddr();
+    unsigned          op1EndOffset = op1Addr->gtFldOffset + genTypeSize(op1Type);
+    unsigned          op2Offset    = op2Addr->gtFldOffset;
+    if ((op1Type == op2Type) && (op1EndOffset == op2Offset) && op1Addr->IsInstance() && op2Addr->IsInstance() &&
+        GenTree::Compare(op1Addr->GetFldObj(), op2Addr->GetFldObj()))
     {
         return true;
     }
```

Comment:
What if `op1Addr` and `op2Addr` are calls to the same function?

Reply:
Good point. This is too permissive, I will revert it back to the original logic.